### PR TITLE
Add origin matcher for wildcard matching

### PR DESCRIPTION
**Pull request by barryvdh**
This allows to check wildcards in origins (subdomains, ports etc)

### CODE REVIEW

File: src/Asm89/Stack/CorsService.php [hunk 1]
```diff
@@ -174,9 +174,16 @@ private function checkOrigin(Request $request)
             // allow all '*' flag
             return true;
         }
+
         $origin = $request->headers->get('Origin');
 
-        return in_array($origin, $this->options['allowedOrigins']);
```
Comment (asm89) [inline src/Asm89/Stack/CorsService.php]: I'd rather keep this the default, but I can see a case for people wanting to use the more involved `OriginMatcher`. How about adding a new optional constructor argument `$originChecker` and a new interface `interface OriginChecker { function check(string $origin): bool; }`. If `$this->originChecker` is `null` we'll use the current codepath. If it's set to something custom we'll use that instead.

Comment (barryvdh) [inline src/Asm89/Stack/CorsService.php]: What are you worried about, breaking changes or slower execution?

**Comment by barryvdh** [comment]
FYI, how other libraries do it: mostly with Regex. But this matcher is much cleaner imho.

https://github.com/expressjs/cors -> accepts strings, regexp and callback
https://github.com/nelmio/NelmioCorsBundle -> Has option for `origin_regex`
https://github.com/ottoyiu/django-cors-headers -> Regular expressions
https://github.com/cyu/rack-cors -> Regular expressions

Alternative could be to allow for an `allowedOriginsMatch` option or something (and perhaps even allowedOriginsRegex` later on. Which would make this path optional, but still easy to use.

File: src/Asm89/Stack/CorsService.php (hunk 1 above)
Comment (asm89) [inline src/Asm89/Stack/CorsService.php]: Both. The current model is just an explicit list of allowed origins. It seems to me it cannot get much simpler. :)

I do understand that this is most likely one of the more important pieces of configuration where we need to be open for extension, hence the suggestion.

**Comment by dawehner** [comment]
```FYI, how other libraries do it: mostly with Regex. But this matcher is much cleaner imho.
```
Personally I would argue that given all those other frameworks had good experiences with a regex, maybe following is not the worst. +1 to have a dedicated option for it, no matter with solution we go with.

**Comment by barryvdh** [comment]
I guess regex and callback would be pretty simple to check

**Comment by sebas5384** [comment]
@barryvdh are this PR updated with the master branch? because we are using this PR as a patch (https://github.com/contentacms/contenta_jsonapi/issues/199), and we can't apply it anymore.

**Comment by barryvdh** [comment]
Fixed with #45

**Pull request closed without merging**